packer/table_transformations/patient_diagnosis_biosource_biomaterial.py

```diff
@@ -47,77 +48,66 @@ def from_obs_df_to_pdbb_df(obs):
-                          (data[data.columns[2]].isnull()))
-        elif column_index == 1:
-            conditions = ((data[data.columns[0]] == row[0]) &
-                          (data[data.columns[1]].isnull()))
-        else:
-            return
-        data.drop(data[conditions].index, inplace=True)
-
-
-def limit_rows_to_lowest_level(data, id_columns):
-    logger.info('Removing redundant rows...')
-    lowest_level_col_id = len(id_columns) - 1
-    for idx, column in enumerate(reversed(id_columns)):
-        if idx == lowest_level_col_id:
+def merge_redundant_rows(data, id_columns):
+    if data.empty:
+        return
+    # sort rows by identifying columns, merging of rows strongly depends on sorting
+    rows = data.sort_values(id_columns, na_position='last').to_dict('records')
+    result_rows = [rows[0]]
+    for row in rows[1:]:
+        row_copied = False
+        for result_row in reversed(result_rows):
+            if _is_ancestor_row(row, result_row, id_columns):
+                _copy_missing_value_to_descendant_row(row, result_row, id_columns)
+                row_copied = True
+            else:
+                break
+        if not row_copied:
+            result_rows.append(row)
+    return pd.DataFrame(result_rows)
+
+
+def _is_ancestor_row(ancestor_row_candidate, descendant_row_candidate, id_columns):
+    for id_column in id_columns:
+        if pd.isna(ancestor_row_candidate[id_column]):
             break
-        drop_higher_level(column, data)
+        if ancestor_row_candidate[id_column] != descendant_row_candidate[id_column]:
+            return False
+    return True
 
 
-def merge_redundant_rows(data, id_columns):
-    # sort rows by identifying columns, merging of rows strongly depends on sorting
-    data.sort_values(id_columns, na_position='first', inplace=True)
-    grouped_data = data.groupby(id_columns[0])
-    # propagate data to lower levels
-    ffill_data = grouped_data.ffill()
-    # limit rows to the lowest level
-    limit_rows_to_lowest_level(ffill_data, id_columns)
-    # drop rows with duplicated identifying columns, keep only the last one
-    ffill_data.drop_duplicates(subset=id_columns, keep='last', inplace=True)
-    return ffill_data
+def _copy_missing_value_to_descendant_row(ancestor_row, descendant_row, id_columns):
+    for column, value in ancestor_row.items():
+        if column in id_columns or pd.isna(value):
+            continue
+        if column not in descendant_row or pd.isna(descendant_row[column]):
+            descendant_row[column] = ancestor_row[column]
 
 
 def update_datatypes(data):
     for col in data.columns:
-        # update integer fields - downcasted to float by ffill function
-        # (NaN does not have an integer representation)
-        data[col] = data[col].apply(to_int)
         # update datetime fields
-        if re.match(r'^[0-9]+\.\sDate', col, flags=re.IGNORECASE):
+        if re.match(r'.*[^\\]*\bdate\b[^\\]*\\$', col, flags=re.IGNORECASE):
```

Comment:
Why do you want to change this regex? the new one does not seem to match e.g. `01. date of biomaterial`

Reply:
https://regex101.com/